catalog/app/containers/Bucket/FallbackToDir.tsx

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react'
+import * as RRDom from 'react-router-dom'
+
+import Placeholder from 'components/Placeholder'
+import * as Model from 'model'
+import * as AWS from 'utils/AWS'
+import * as NamedRoutes from 'utils/NamedRoutes'
+import * as s3paths from 'utils/s3paths'
+
+import * as requests from './requests'
+import { displayError } from './errors'
+
+const Loading = Symbol('loading')
+
+const Dir = Symbol('dir')
+
+const File = Symbol('file')
+
+// If object exists, then this is 100% an object page
+function useIsObject(handle: Model.S3.S3ObjectLocation) {
+  const s3 = AWS.S3.use()
+  const [exists, setExists] = React.useState<typeof Loading | boolean | Error>(Loading)
+
+  React.useEffect(() => {
+    const { bucket, key, version } = handle
+    requests
+      .getObjectExistence({
+        s3,
+        bucket,
+        key,
+        version,
+      })
+      .then(
+        requests.ObjectExistence.case({
+          Exists: () => true,
+          _: () => false,
+        }),
+      )
+      .then(setExists)
+      .catch(setExists)
+  }, [s3, handle])
+
+  return exists
+}
+
+// If prefix contains at least something, then it is a directory.
+// S3 can not have empty directories, because directories are virtual,
+//    they are based on paths for existing keys (file)
+function useIsDirectory(handle: Model.S3.S3ObjectLocation, pause: boolean) {
+  const bucketListing = requests.useBucketListing()
+  const [isDir, setIsDir] = React.useState<typeof Loading | boolean | Error>(Loading)
+
+  React.useEffect(() => {
+    if (pause) return
```

Comment:
it's probably worth it to reset state here as well (before `if (pause) ...`)

Reply:
Added Loading state